File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.PriorityQueue;
import java.util.Arrays;

class Solution {
    public int solution(String[][] book_time) {
        int answer = 0;
        
        Arrays.sort(book_time, (o1,o2) -> {
            int s1 = convertTime(o1[0]);
            int s2 = convertTime(o2[0]);
            if(s1 == s2) {
                return Integer.compare(convertTime(o1[1]), convertTime(o2[1]));
            }
            return Integer.compare(s1,s2);
        });
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for(String[] book : book_time){
            int start = convertTime(book[0]);
            int end = convertTime(book[1]);
            if(!pq.isEmpty() && start >= pq.peek() + 10){
                pq.poll();
            }   
            pq.offer(end);
        }
        return pq.size();
    }
    
    private int convertTime(String time){
        return Integer.parseInt(time.split(":")[0]) * 60 + Integer.parseInt(time.split(":")[1]);
    }
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.kt
import java.util.*

class Solution {
    
    fun solution(book_time: Array<Array<String>>): Int {
        var answer: Int = 0
        val pq = PriorityQueue<Reservation>(compareBy<Reservation> { it.transStart }.thenBy { it.transEnd })
        pq.addAll(book_time.map { Reservation(it[0], it[1]) }.toList())

        val room: MutableList<Int> = mutableListOf()

        while (pq.isNotEmpty()) {
            val now = pq.poll()
            if (room.size == 0) {
                room.add(now.transEnd)
                continue
            }
            var flag : Boolean = false
            for (index in room.indices) {
                if (now.transStart >= room[index] + 10) {
                    room.removeAt(index)
                    room.add(now.transEnd)
                    flag = true
                    break
                }
            }
            
            if(!flag) {
                room.add(now.transEnd)
            }
        }
        
        return room.size
    }

    data class Reservation(val start: String, val end: String){
        val transStart: Int = convertTimeType(start.split(":"))
        val transEnd: Int = convertTimeType(end.split(":"))
        private fun convertTimeType(time: List<String>) : Int {
            return ((time[0].toInt()) * 60) + (time[1].toInt())
        }
    }
}